### ISSUE TITLE
WG New Spec: Scroll-Triggered Animations

### ISSUE THREAD

**Issue by DavMila** [w3ctag/design-reviews#1167]
### Specification

https://drafts.csswg.org/css-animations-2/#timeline-triggers

### Explainer

https://github.com/explainers-by-googlers/scroll-triggered-animations/blob/main/README.md

### Links

- The WG's request for this TAG review: None
- Previous early design review, if any: None (although https://github.com/w3ctag/design-reviews/issues/1111 is related)
- An introduction to the feature, aimed at unfamiliar audiences: https://github.com/explainers-by-googlers/scroll-triggered-animations/blob/main/README.md#proposed-solution  
- A description of the problems that end-users were facing before this proposal:  https://github.com/explainers-by-googlers/scroll-triggered-animations/blob/main/README.md#introduction 
- Alternatives considered: https://github.com/explainers-by-googlers/scroll-triggered-animations/blob/main/README.md#considered-alternatives  
- Examples of how to use the proposal to solve the end-users' problems: https://github.com/explainers-by-googlers/scroll-triggered-animations/blob/main/README.md#example 
- What do the end-users experience with this proposal: Seamless scroll-triggered animations  
- User research you did to validate the problem and/or design, if any: None
- Web Platform Tests:  https://wpt.fyi/results/scroll-animations/animation-trigger?label=experimental&label=master&aligned 

### The specification

- [x] Follows the [Web Platform Design Principles](https://www.w3.org/TR/design-principles/).
- [x] Includes Security and Privacy Considerations sections based on answers to the [Security/Privacy Questionnaire](https://www.w3.org/TR/security-privacy-questionnaire/).

### Where and by whom is the work is being done?

- GitHub repo: None
- Primary contacts:
  -  @DavMila, Google, CSSWG participant
  -  @flackr, Google, CSSWG participant
  -  @tabatkins, Google, CSSWG participant, spec editor
- Organization/project driving the specification: Google
- This work is being funded by: Google
- Primary standards group developing this feature: CSSWG
- Incubation and standards groups that have discussed the design:
  - CSSWG 
    - [#12652](https://github.com/w3c/csswg-drafts/issues/12652#issuecomment-3329841437)
    - [#12611](https://github.com/w3c/csswg-drafts/issues/12611#issuecomment-3382276301)
    - [#12336](https://github.com/w3c/csswg-drafts/issues/12336#issuecomment-3206545734)
    - [#12119](https://github.com/w3c/csswg-drafts/issues/12119#issuecomment-2985057242) 

### Feedback so far

- Active horizontal reviews: We filed csswg-draft issue [#13010](https://github.com/w3c/csswg-drafts/issues/13010) to solicit feedback from the broader CSSWG on the overall spec text. 
- Multi-stakeholder feedback:
  - Chromium comments: Chromium has an experimental prototype of this feature.
  - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1315 
  - WebKit comments: https://github.com/WebKit/standards-positions/issues/576
- Major unresolved issues with or opposition to this specification: None
- Status/issue trackers for implementations: 
  - [Chromestatus](https://chromestatus.com/feature/5181996801982464)

### You should also know that...

There is also a [Web Animations](https://drafts.csswg.org/web-animations-2/) component to this proposal but only the [CSS Animations](https://drafts.csswg.org/css-animations-2/#animation-triggers) spec is up to date. We plan to update the Web Animations spec after the CSSAnimations spec has been reviewed in [#13010](https://github.com/w3c/csswg-drafts/issues/13010). We expect the web animations API to reflect the css animations API.

---

Track conversations at https://tag-github-bot.w3.org/gh/w3ctag/design-reviews/1167

**Comment by jyasskin** [comment]
We have a week before the Blink folks will approve this Intent to Ship: https://groups.google.com/a/chromium.org/g/blink-dev/c/TZLFLjzt4II/m/nzUrqB4CAwAJ

**Comment by matatk** [comment]
Hi @DavMila, thanks for your review request.

We appreciate this is an increasingly used pattern, and it's important to help authors to get it right. On those lines, we have several concerns:

* The robustness of this pattern on sites in general - e.g. when afforded by scripting - can be poor. For example: triggers firing the wrong number of times; things going wrong if the user scrolls back. Have you considered how these issues may be resolved through standardization?

* Historically, all page content was loaded in and was available to ATs - it might just be off screen, but could be explored. If content is loaded dynamically in response to scrolling, this could exclude some users. (It doesn't seem like that's _inherent_ here, but it seems like a pattern that may more prominently emerge following this.)

* Scrolling and animation can cause huge cognitive issues (e.g. distracting, disorienting). We have `prefers-reduced-motion` but we're not sure how much consideration has been given to providing guidance for authors on how to ensure an equitable UX - especially in cases of complex animations revealing - or even themselves acting as - content.

* Does this interact with/affect the focus in any way?

There seems to be justification for an 'accessibility considerations' section, assuming ways to address these concerns are known, or could be found. APA WG would be happy to help by reviewing any content you might create for such a section.

We look forward to hearing your thoughts on the above.

**Comment by matatk** [comment]
Hi @DavMila, just wondering if you had any thoughts in response to our comment above?

**Comment by DavMila** [comment]
> Hi [@DavMila](https://github.com/DavMila), just wondering if you had any thoughts in response to our comment above?

Hi @matatk, we discussed this shortly after your comment but I apologize for taking so long in getting back to this.

> * The robustness of this pattern on sites in general - e.g. when afforded by scripting - can be poor. For example: triggers firing the wrong number of times; things going wrong if the user scrolls back. Have you considered how these issues may be resolved through standardization?

One of the choices we made was to leverage [scroll timelines](https://developer.mozilla.org/en-US/docs/Web/API/ScrollTimeline) which were introduced along with [scroll-driven animations](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Scroll-driven_animations). Timelines, via the [ranges](https://drafts.csswg.org/scroll-animations/#view-timelines-ranges) API, offer developers a robust way of referring to portions of a scroll container. To use the “user scrolls back” case mentioned as an example, timeline triggers are designed to allow authors specify two ranges - this means that for a given side of the scroll port, authors can separate the boundary at which scrolling forward fires the animation from the boundary at which scrolling backwards does. One of the use cases we prioritized simplifying was being able to specify the ranges as `contain / cover` which provides a separation of the trigger points so that on one end of the scroll port triggering only happens when the element comes fully into view or goes fully out of view.
We believe that this boundary separation, together with semantically connected “entry” and “exit” concepts (i.e. entry and exit must alternate) are likely to foster the development of UIs that are predictable and user-friendly.

> * Historically, all page content was loaded in and was available to ATs - it might just be off screen, but could be explored. If content is loaded dynamically in response to scrolling, this could exclude some users. (It doesn't seem like that's _inherent_ here, but it seems like a pattern that may more prominently emerge following this.)

I think our expectation is that use of this feature will apply to elements coming in and out of the viewport/scrollport, which are typically elements that are already loaded into the DOM.

> * Scrolling and animation can cause huge cognitive issues (e.g. distracting, disorienting). We have `prefers-reduced-motion` but we're not sure how much consideration has been given to providing guidance for authors on how to ensure an equitable UX - especially in cases of complex animations revealing - or even themselves acting as - content.

We recognize that this is an important issue to consider and we think that this is something that developers should be considering in the context of animations in general. To this end, we noticed that the [css-animation spec](https://drafts.csswg.org/css-animations-2/) doesn’t have an accessibility section similar to the [css-transitions spec](https://www.w3.org/TR/css-transitions-1/#accessibility). So, we are [working on](https://github.com/w3c/csswg-drafts/pull/13243) adding an accessibility section with a strong recommendation on prefers-reduced-motion. I think that one potential advantage of declarative scroll-triggered animations is the ability for the user agent to make modifications to the animations on a page according to the user’s needs. It might be more difficult for a user agent to do this with scripted scroll-triggered animations.

> * Does this interact with/affect the focus in any way?
> 

No, it doesn’t directly affect focus. In theory, it would be possible for an author to set up an animation that affects other CSS properties that, in turn, interact with focus , e.g. display. However, that’s true of animations today.

**Comment by matatk** [comment]
Hi @DavMila, thanks for your helpful feedback, and sorry for the slow reply.

We have a couple of questions; starting with the more specific...

1. Regarding the use of `contain` and `cover`: we think this is going to cause developer confusion - whilst re-using syntax is generally a good thing, it seems that the meaning of the keywords here diverges from how they're used in CSS in general. Has the group discussed alternative names - if they were rejected, could you point us to the rationale?

2. Should the UA have some control over the speed of these animations in particular? Regarding the issue that they could be distracting for people, or simply happen too fast for some people to process (both of these could affect anyone who is distracted, or has specific cognitive disabilities).

   As these are a 'new' type of animation (from the spec's perspective), perhaps there's an opportunity to consider whether we could give more agency to the UA in how they are rendered?

   Specifically, if there's an opportunity here to make steps towards [giving users more control over motion](https://github.com/w3ctag/gaps/issues/15), it would be good to do so.

With respect to your [proposed accessibility considerations section](https://github.com/w3c/csswg-drafts/pull/13243/changes), this looks great so far; thanks. APA WG started drafting one, but since I saw you had been working on this, and it covers much of what APA wanted to say, we paused that, and will check it out when the PR is merged.

**Comment by DavMila** [comment]
Thanks for the questions

> 1. Regarding the use of `contain` and `cover`: we think this is going to cause developer confusion - whilst re-using syntax is generally a good thing, it seems that the meaning of the keywords here diverges from how they're used in CSS in general. Has the group discussed alternative names - if they were rejected, could you point us to the rationale?

Oh, valid point. I agree that the meaning isn't quite the same as with properties like `object-fit` and `background-size` which I just looked up. However I would point out that this feature doesn't actually introduce this interpretation of those keywords or the use of these keywords in the context of timelines. It is only adhering to what was adopted when timelines were introduced in [scroll-driven animations](https://github.com/w3ctag/design-reviews/issues/828). In addition, I don't think we've received feedback, since the launch of SDA, suggesting this is an issue for developers, so perhaps the context is different enough from the other properties that it mitigates the fact that it is a different use of the keywords, which I agree with.
  
> 2. Should the UA have some control over the speed of these animations in particular? Regarding the issue that they could be distracting for people, or simply happen too fast for some people to process (both of these could affect anyone who is distracted, or has specific cognitive disabilities).
>    As these are a 'new' type of animation (from the spec's perspective), perhaps there's an opportunity to consider whether we could give more agency to the UA in how they are rendered?
>    Specifically, if there's an opportunity here to make steps towards [giving users more control over motion](https://github.com/w3ctag/gaps/issues/15), it would be good to do so.

I think it could be a good idea for the UA to control over the speed of these animations. This feature may actually be a positive step towards that possibility from the point of view that prior to this feature, even though scroll-triggered animations exist, user agents have no concrete concept of what a "scroll-triggered animation" is because it's probably all carried out by script in various different ways. So you could have no reliable way to offer a user the option to "disable all scroll-triggered animations." With this feature I think it becomes more of a possibility, and I think this feature could help with exploring good options for what UA control over scroll-triggered animations should look like.

**Comment by astearns** [comment]
> 1. Regarding the use of `contain` and `cover`: we think this is going to cause developer confusion - whilst re-using syntax is generally a good thing, it seems that the meaning of the keywords here diverges from how they're used in CSS in general. Has the group discussed alternative names - if they were rejected, could you point us to the rationale?

This issue discussion is the earliest bit I’ve found for cover/contain in this context: https://github.com/w3c/csswg-drafts/issues/6674

**Comment by matatk** [comment]
Thanks @DavMila and @astearns for your review request and further info. We are happy to see the work continue, and to close this review.

However, whilst we appreciate the importance of consistency with respect to `contain` and `cover`, having been introduced in scroll-driven animations, and in general we agree that the terms are likely not universally confusing, we are concerned about how CSS (and other groups) can get reliable input from developers on what they find confusing, as specs evolve. We think it most likely that, particularly with minor-to-moderate things that confuse developers, most _aren't_ likely to report those issues as bugs on the spec.

This is of course a key issue that we all need to work to address - ensuring we're reaching the [second-highest-priority constituency [link for readers new to W3C]](https://www.w3.org/TR/design-principles/#priority-of-constituencies) - though with near-universally deployed specs such as CSS it is of course amplified.

If there's anything you think we could do as a community, perhaps kicking of with a TPAC breakout on how to ensure we're getting feedback as inclusively as possible, please let us know.

Finally, we've recently discovered a case where scroll-driven animations can be used to create a scroll spy that can make network requests based on scroll positioning, even when JavaScript is disabled. This opens up privacy concerns, we believe that the same issue applies to scroll triggered animations and would be interested in the CSSWG's thinking on this.
